src/audio/dai.c

```diff
@@ -222,6 +222,10 @@ static void dai_free(struct comp_dev *dev)
 		dd->chan->dev_data = NULL;
 	}
 
+#if CONFIG_IPC_MAJOR_4
+	dai_trigger(dd->dai, COMP_TRIGGER_STOP, dev->direction);
```

Comment:
> so looks like an incorrect IPC3 sequence could try and free whilst running ?

Seems like it could

```
if (dai_is_active(dd->dai))
    dai_trigger(dd->dai, COMP_TRIGGER_STOP, dev->direction);
```
We could do some kind of "sanity check" there for both IPCs, though we can't only check for active state (f.e. in the case i'm working on it goes PAUSE->RESET->DELETE, which renders the dai in COMP_STATE_READY when trying free the component). 
Looks like checking for `!COMP_STATE_PREPARE`  should be correct, am I right?

Reply:
> Looks like checking for `!COMP_STATE_PREPARE` should be correct, am I right?

@lyakh can you confirm the active states and point @kfrydryx to the state transition diagram. Thanks

Reply:
@lgirdwood do you mean this https://github.com/lyakh/sof/commit/9f4edc49a2a47ba5226b40ae8389d5010f46fc50#diff-d174dfca367ecd89c8e7c345c9a4e580f0fe019d9efd238a42a43d5296a376f7 (comment starting at line 15) my attempt at documenting component state transitions? I'm not sure it's correct or complete. It's based at least partially on https://github.com/thesofproject/sof/blob/main/test/cmocka/src/audio/component/comp_set_state.c and on https://github.com/thesofproject/sof/blob/7b5aa7281370c287a16600c7b010279c3fd81def/src/audio/component.c#L58 but a real complete authoritative table of these transitions should probably be based on ALSA flows and on the kernel driver behaviour. Then again, the firmware probably shouldn't rely on either of those and should check valid transitions itself.

Reply:
@kfrydryx @lyakh so it sound we may need a helper API call like
```c
bool pipeline_is_active(pipeline)
{
    /* return true if the pipeline is active and scheduled to copy data */
}
```

Reply:
We've discussed this a bit offline, and it seems that there's a mismatch between the docs (f.e. [there](https://thesofproject.github.io/latest/architectures/firmware/components/component-overview.html?highlight=states)) and code. In the code, reset operation is always valid and calling it in active/paused state omits the trigger_stop call. In this case I'd like to propose a check, as stated before:
```
if (dev->state != COMP_STATE_PREPARE){
    dai_trigger(dd->dai, COMP_TRIGGER_STOP, dev->direction);
}
```
Which might result in, according to the docs, invalid call for _stop_ from COMP_STATE_READY. But the state chart **is incomplete**, so I think we should make this transition valid.

Alternatively we could internally call _stop_ when reset is invoked in an invalid state (again, invalid based on the docs. Code says reset is always valid). 

@lgirdwood What's Your opinion?

Reply:
Agree, It sounds like the logic would be safer if we were to call reset() as part of invalid state handling (and resent calls STOP etc). reset() should also be called in the free() path.